### DISCUSSION TITLE
recursive version of `just`

### DISCUSSION THREAD

**Discussion by Jasha10** [mit-ll-responsible-ai/hydra-zen#293] [Show and tell]
```python
"""
A few months ago @rsokl shared a vision with me: what if arbitrary python objects could be
serialized to an OmegaConf-compatible format, then modified and composed using Hydra, and finally
reanimated with a call to `instantiate`?

In this literate program, I outline an approach to automatic, recursive conversion of
OmegaConf-incompatible dataclass instances to an OmegaConf-compatible form. This is something like a
poor man's implementation of the auto-config support proposed in hydra-zen issue
https://github.com/mit-ll-responsible-ai/hydra-zen/issues/257 .
"""

#######################
# Motivating use case #
#######################

# Below we create an object `foobar` containing data and type-hints unsupported by Hydra.

from collections.abc import Callable
from dataclasses import dataclass
from typing import Any, TypeAlias

Interface1: TypeAlias = Callable[[int], int]
Interface2: TypeAlias = Callable[[str], str]


@dataclass
class Nested:
    x: Any


@dataclass
class Stuff:
    field1: Interface1
    field2: Interface2
    field3: Nested


def foo(i: int) -> int:
    return i


def bar(s: str) -> str:
    return s


def baz(f: float) -> float:
    return f


foobar = Stuff(foo, bar, Nested(baz))

###############
# The problem #
###############

# The type hints `Interface1` and `Interface2` are not supported by OmegaConf, and neither are the
# values `foo`/`bar`/`baz`. As such, we cannot serialize the object `foobar` using OmegaConf:

from omegaconf import OmegaConf, ValidationError
from pytest import raises

with raises(ValidationError):
    OmegaConf.structured(foobar)

# This means `foobar` cannot pariticipate in config composition via Hydra.

############
# The Goal #
############

# Given `foobar`, how can we create a Hydra-compatible dataclass `BuildsFooBar` such that
# `instantiate(BuildsFooBar) == foobar`?

# Here is a solution using `hydra_zen.builds`:

from hydra.utils import instantiate
from hydra_zen import builds

BuildsFooBar = builds(Stuff, foo, bar, builds(Nested, baz))
assert instantiate(BuildsFooBar) == foobar
assert OmegaConf.to_yaml(BuildsFooBar) == """\
_target_: __main__.Stuff
_args_:
- _target_: hydra_zen.funcs.get_obj
  path: __main__.foo
- _target_: hydra_zen.funcs.get_obj
  path: __main__.bar
- _target_: __main__.Nested
  _args_:
  - _target_: hydra_zen.funcs.get_obj
    path: __main__.baz
"""

# Unlike `foobar` itself, the dataclass `BuildsFooBar` (and its instances) are fully compatible with
# OmegaConf + Hydra.

# How can the above workflow be improved? Ideally, we'd be able to transform the instance `foobar`
# into the class `BuildsFooBar` in a fully-automatic fashion. I envision something like this:

#   BuildsFooBar2 = recursive_just(foobar)
#   assert instantiate(BuildsFooBar2) == foobar

# The idea is for `recursive_just(obj)` to visit sub-objects of `obj`, converting each of them
# to be Hydra-compatible. Like the non-recursive `hydra_zen.just` function, this proposed "recursive
# just" operator is idempotent.


############################
# Prototype implementation #
############################

from dataclasses import fields, is_dataclass
from hydra_zen import just


def recursive_just(obj):
    if is_dataclass(obj) and not isinstance(obj, type):
        # obj is a dataclass instance
        converted_fields = {}
        for field in fields(obj):
            value = getattr(obj, field.name)
            converted_fields[field.name] = recursive_just(value)
        return builds(type(obj), **converted_fields)

    else:
        return just(obj)


BuildsFooBar2 = recursive_just(foobar)
assert instantiate(BuildsFooBar2) == foobar
assert OmegaConf.to_yaml(BuildsFooBar2) == """\
_target_: __main__.Stuff
field1:
  _target_: hydra_zen.funcs.get_obj
  path: __main__.foo
field2:
  _target_: hydra_zen.funcs.get_obj
  path: __main__.bar
field3:
  _target_: __main__.Nested
  x:
    _target_: hydra_zen.funcs.get_obj
    path: __main__.baz
"""
```

cc @pixelb @jieru-hu

Edit: fixed a bug above

**Comment by rsokl** [comment]
Hi @Jasha10, thanks so much for taking the time to write this up! I like `recursive_just` and think that it would be a natural fit for hydra-zen.

I am currently playing around with your proposed implementation and am working on a write-up. I wanted to post in the meantime to let you know that I am working on this. I am teaching an intensive course this month, which is pulling me away from hydra-zen for the time being, so my responses may be more delayed than I would like them to be 😅 

Thanks again for sharing this idea!

**Comment by addisonklinke** [comment]
Thanks for sharing @Jasha10! Would this be related to https://github.com/mit-ll-responsible-ai/hydra-zen/issues/257 with the addition of recursion?

**Reply by Jasha10** [reply]
> Would this be related to #257

It would indeed!

**Comment by rsokl** [comment]
## Response

Hi @Jasha10 ! I am finished teaching and am back in the swing of things at work! Thanks for your patience and understanding for the past few weeks.

## A Brief Review of This Proposal

Let me first summarize my understanding of the novel functionality that is being proposed here. Wherease `builds` and `just` are both used to construct a config that describes a single object in one's program, `recursive_just` is meant to operate on hierarchical/nested structured configs. Specifically one to can write a hierarchical structured config without paying heed to Hydra's [^1] limited support for various type annotations[^2] and values[^3], and `recursive_just` will automatically create a Hydra-compatible form for it.

[^1]: For simplicity's sake, I am going to ignore the distinction between Hydra and omegaconf here, and just use Hydra as a catch-all.

[^2]: Because the "sanitized" config will instantiate to an instance of the original dataclass type, users can rely on this post-instantiation type information within their task function. 

[^3]: Values of types that [have registered support with hydra-zen](https://mit-ll-responsible-ai.github.io/hydra-zen/api_reference.html#configuration-value-types-supported-by-hydra-and-hydra-zen), e.g. `complex` numbers, will be automatically supported as well. 


To demonstrate the utility of this in practice, let's design a structured config using a pydantic-based dataclass, which we can make compatible with Hydra while retaining the runtime type-checking and rich types associated with our original dataclass.


```python
import dataclasses
from typing import List, Optional, Literal

from pydantic.dataclasses import dataclass as pydantic_dataclass
from pydantic import PositiveInt


@pydantic_dataclass
class User:
    name: str
    age: PositiveInt  # typically Hydra users would have to use `int` here


@pydantic_dataclass
class Match:
    user_a: User = User(name="Bob", age=25)
    user_b: User = User(name="Alice", age=27)
```

Now we can use `recursive_just` to make an Hydra-compatible config from this pydantic-based dataclass, and we still benefit from our rich type annotations and pydrantic's runtime type-checking via instantiation:  

```python
from hydra.errors import InstantiationException
conf = recursive_just((User(name="Bob", age=20)))

# Attempting to overwrite age with a negative number..
with raises(InstantiationException):
    # Pydantic's runtime type-checking raises an error during the instantiation process.
    instantiate(conf, age=-10)  # error: age < 0

profile: User = instantiate(conf, age=11)
assert isinstance(profile, User)
assert profile == User("Bob", 11)
```


And here is the recursion in action:

```python
from hydra_zen import to_yaml
sanitized_match = recursive_just(Match())

assert to_yaml(sanitized_match) == """\
_target_: __main__.Match
user_a:
  _target_: __main__.User
  name: Bob
  age: 25
user_b:
  _target_: __main__.User
  name: Alice
  age: 27
"""

assert instantiate(sanitized_match) == Match()
```

I like this proposal a lot! I think that this reduces the barrier to entry for new Hydra users while also enabling advanced patterns (e.g., the demonstrated cross-compatibility with pydantic!). The improved ergonomics afforded by the recursion is also a big win, and the gains here will grow as hydra-zen's [auto-config support](https://github.com/mit-ll-responsible-ai/hydra-zen/issues/257) continues to improve (e.g., see https://github.com/mit-ll-responsible-ai/hydra-zen/issues/294).

## Implementation Details

One change to this prototype that I have identified thus far is that I would like to support dataclass-types in addition to the proposed dataclass-instances; this is motivated by the fact that Hydra largely treats dataclass types and instances on an equal footing.
Here is a modified prototype that incorporates this change (as well as enforcing a different condition for idempotence, which I'll discuss below):

```python
from hydra_zen import just, builds, to_yaml
from hydra_zen.structured_configs._type_guards import is_builds
from dataclasses import is_dataclass, fields

def recursive_just(obj):
    if is_dataclass(obj) and not is_builds(obj):
        # obj is a dataclass type or instance that does not have a _target_ field
        type_ = obj if isinstance(obj, type) else type(obj)

        converted_fields = {}

        for field in fields(obj):
            if field.init and hasattr(obj, field.name):
                value = getattr(obj, field.name)
                converted_fields[field.name] = recursive_just(value)
        Conf = builds(
            type_,
            populate_full_signature=True,
            **converted_fields
        )
        return Conf if type_ is obj else Conf()

    else:
        return just(obj)
```

Thus `recursive_just` could be applied directly to the `User` pydanatic-dataclass type  [^4].

```python
# support for dataclass types (in addition to instances)
assert to_yaml(recursive_just(User)) == """\
_target_: __main__.User
name: ???
age: ???
"""

# `recursive_just(User)` returns dataclass type
assert isinstance(recursive_just(User), type)  # Builds_User

# `recursive_just(User("bob", 12))` returns dataclass instance
assert not isinstance(recursive_just(User("bob", 12)), type)  # Builds_User("bob", 12)  
```

This is effectively identical to `builds(User, populate_full_signature=True)`, but it also includes recursion over fields that store nested structured configs, which is what makes the functionality unique. 

Supporting dataclass types impacts the nice property of idempotence that Jasha's implementation possesses. To remedy this, `recursive_just` will only apply to dataclasses that do not have a `_target_` field. This restriction prevents people from accidentally creating a structured config that instantiates to a structured config that instantiates to a structured config... and so on.

```python
from dataclasses import dataclass

# demonstrating idempotence

compat_user = recursive_just(User("bob", 12))

assert hasattr(compat_user, "_target_")
assert recursive_just(compat_user) is compat_user

@dataclass
class HasTarget:
    _target_: str = "hi"

conf = HasTarget()
assert recursive_just(conf) is conf
```

[^4]: One thought that I had is: is there value to preserving basic parent/child parity through `recursive_just`?:

    ```python
    assert isinstance(User(age=10), User)
    
    CompatUser = recursive_just(User)
    compat_user = recursive_just(User(age=10))  # creates an instance of `CompatUser`
    assert isinstance(compat_user, CompatUser)
    ```

    This would add substantial complications to the implementation, so I'd prefer to forego this unless there is a truly pressing benefit.


## The API

**note: I ended up changing course on the API design that I laid out here, see https://github.com/mit-ll-responsible-ai/hydra-zen/discussions/293#discussioncomment-3503404 for revised details**

The name `recursive_just` was appropriate when this functionality was only applied to dataclass instances. Given that we will support dataclass types as well and given that the name `just` is not a very clear description of what is going on here, I think a name like `create_hydra_compat_config` is appropriate (I am very open to alternative suggestions!).

We will also need to allow users to pipe through 

```python
def create_hydra_compat_config(
                    config: Union[Dataclass, Type[Dataclass]], 
                    hydra_convert=None,  # Sent to all recursive calls of `create_hydra_compat_config`
                    
                    # these only apply to the resulting top-level dataclass
                    dataclass_name: Optional[str] = None,
                    bases: Tuple[Type[Dataclass], ...] = ()
    )
    -> Builds[Type[Dataclass]]:
    """Given a dataclass type or instance, constructs a Hydra-compatible structured config that, 
    when instantiated, returns an instance of the original dataclass. This applies recursively to 
    nested dataclasses"""
    ...

```

I am hoping that the name conveys to the user that this function is meant to operate specifically on structured configs / dataclasses, and that a new, Hydra-compatible, structured config is returned.

## Next steps

I would love to get general feedback on this (including possible alternative names to `create_hydra_compat_config`. In the meantime, I will begin working on an implementation with tests to help ferret out any issues with the API and/or implementation details that I am failing to foresee.

**Reply by Jasha10** [reply]
> I would like to support dataclass-types in addition to the proposed dataclass-instances

Sounds good!

---

> One thought that I had is: is there value to preserving basic parent/child parity through `recursive_just`?

The value I see is related to the restrictions that `OmegaConf.merge` imposes on parent/child structured config classes. When it's time to merge configs during hydra's composition phase, having proper parent/child relationships would enable merging values that come from different calls to `recursive_just`, e.g:
```python
cfg1 = recursive_just(User)(name="???", age=7)
cfg2 = recursive_just(User)(name="Bond", age="???")
assert OmegaConf.merge(cfg1, cfg2) == {name: "Bond", age: 7}  # raises ValidationError
```
The above call to `OmegaConf.merge` will fail with a `ValidationError` unless we can somehow
resolve the inequality `recursive_just(User) != recursive_just(User)`, which
of course results from calling the `builds` function twice:
```python
>>> assert builds(User) != builds(User)
>>> OmegaConf.merge(builds(User), builds(User))
Traceback (most recent call last):
  ...
omegaconf.errors.ValidationError: Merge error: Builds_User is not a subclass of Builds_User. value: {'_target_': '__main__.User'}
  ...
```
OmegaConf uses `issubclass` to decide whether structured configs can legally be merged. Since `issubclass(builds(User), builds(User))` is False, the merge fails.

No doubt you're right about the implementation complexity that fixing this would introduce. *If* you decide to pursue this direction (of preserving parent/child parity through `recursive_just`, or potentially through `builds`), I'd be happy to brainstorm with you on potential solutions.

---

> I am hoping that the name conveys to the user that this function is meant to operate specifically on structured configs / dataclasses, and that a new, Hydra-compatible, structured config is returned.

For the record, I chose the name `recursive_just` thinking that would be cool to write `just(..., recursive=True)`. I agree that `recursive_just` is less fitting now that the proposed functionality extends beyond dataclass instances to dataclass types.

My personal opinion is that the name `create_hydra_compat_config` is a bit long and hard to remember when compared with the other config-creation functions exposed by hydra-zen (`make_config`, `builds`, `just`, `hydrated_dataclass`). I think just `create_compat_config` would be an improvement. Here are few more brainstormed ideas:
- `autoconf`
- `compat_just`
- `convert_config`

**Reply by rsokl** [reply]
> For the record, I chose the name `recursive_just` thinking that would be cool to write `just(..., recursive=True)`. I agree that recursive_just

Yep I totally got that and really like how you designed this proposal to fit so cleanly with `just` 😄 . Because of this, I have been/am very tempted to make `just(..., recursive=True)` and perhaps `builds(..., populate_full_signature="recursive")` the ultimate API for handling dataclass instances and types, respectively. I keep going back and forth on this in my mind...


>  If you decide to pursue this direction (of preserving parent/child parity through recursive_just, or potentially through builds), I'd be happy to brainstorm with you on potential solutions.

Great! Much appreciated. I will likely get a working version of your proposal running without paying heed to this parent/child issue at first. The preliminary implementation will be sufficiently simple (not much more complicated than supporting, e.g., [auto-config for complex numbers](https://github.com/mit-ll-responsible-ai/hydra-zen/blob/7675970ed73cfb9d83fac513ad8c7cc972da184a/src/hydra_zen/structured_configs/_value_conversion.py#L18-L29)) that it will not be hard for me to go back and do a substantial re-imagining of it later if need be.

https://github.com/mit-ll-responsible-ai/hydra-zen/pull/299 was the main blocker on this, which is now resolved  so I hope to make some progress on recursive-just/builds this week!

**Reply by addisonklinke** [reply]
@rsokl is there a reason we need to initialize a `Match` object as the parameter to `recursive_just()` rather than just passing the class definition? It seems like we should be able to determine the default field values and types from the type hints like with `builds()`, right?

```python
sanitized_match = recursive_just(Match())
```

> this function is meant to operate specifically on structured configs / dataclasses

Is there a reason it couldn't support normal classes? If the class is not already a dataclass instance, run `builds()` first to ensure it is before continuing. My guess is I'm missing a key detail here 

As for the API, it is getting difficult for me (as someone who's less familiar with the Hydra internals) to keep the differences straight between `builds`, `just`, `like`, and this new `recursive_just`. A side-by-side summary of canonical use cases would be pretty helpful documentation for users to understand which is appropriate for their situation. For now, I've been trying to track the discussion across various issue threads

**Reply by Jasha10** [reply]
@addisonklinke:

> is there a reason we need to initialize a Match object as the parameter to `recursive_just()` rather than just passing the class definition?

Working with Ryan's implementation of `recursive_just`, we have
`to_yaml(recursive_just(Match())) == to_yaml(recursive_just(Match))`,
so the two are equivalent in this case.

> it is getting difficult for me (as someone who's less familiar with the Hydra internals) to keep the differences straight between `builds`, `just`, `like`, and this new `recursive_just`

All of these functions create instantiable configs; I think of them in terms of how they compose with `instantiate`:

- `builds` usually satisfies `instantiate(builds(C, x, y)) ≈ C(x, y)` for callable objects `C`.
  From this perspective, `instantiate(builds(C, x, y))` looks a lot like `partial(C, x, y)()`.
    Calls can be nested: `instantiate(builds(C, x, builds(D, y))) ≈ C(x, D(y))`.
- `just` is like a partial right inverse to `instantiate`: `instantiate(just(X)) ≈ X`,
  though there is special treatment if X is a dataclass / dataclass instance.
- `recursive_just` is like `just`, with some support for recursion to subobjects:
    ```python
    instantiate(recursive_just(C(x, D(y)))) ≈ C(x, D(y))
    ```
    (here `C` and `D` are assumed to be a dataclasses that do not have a `_target_` field)
- `make_config` satisfies `instantiate(make_config("x", y=z)) ≈ {"x": "???", "y": z}`
- I'm not familiar with `like`.

> Is there a reason it couldn't support normal classes?

Class types (but not class instances) should work out of the box:
```python
class Foo:
    ...
assert instantiate(recursive_just(Foo)) is Foo
```

Handling instances of classes would be more complicated. I'm guessing such functionality is out of scope at the moment given the direction Ryan is going with this. Given an instance (e.g. `foo = Foo("bar", "baz")`), `recursive_just` would need to create a config such that `instantiate(cfg) ≈ foo`. For the general case, I think this would require the user to supply some logic that decomposes the `Foo` instance into a suitable config. This is why I mentioned issue #257 in the original post's docstring; #257 proposes a registry for user-defined conversion logic (though the use-case there is somewhat different).

**Reply by rsokl** [reply]
Thanks for this fantastic summary of our API, @Jasha10 . This is all spot-on, including your allusions to #257 . I also think about `just` and `builds` in terms of how they compose under `instantiate`. I hope to add a section to the Explanations portion of our docs that provides a summary that is similar to this.

> I'm not familiar with `like`

I only ever made a rough prototype of `like`. [It was designed to use a config to record interaction with an object that would get replayed via instantiation](https://github.com/mit-ll-responsible-ai/hydra-zen/issues/205#issuecomment-1030189425). @Jasha10  you had [pointed out](https://github.com/mit-ll-responsible-ai/hydra-zen/issues/205#issuecomment-1030305336) that this was similar to so `unittest.mock` functionality. I currently do not have plans to move forward with that functionality.


A quick update on recursive-just: I have a reasonably well-tested implementation of this functionality here: https://github.com/mit-ll-responsible-ai/hydra-zen/pull/301 . I ultimately am walking back some of the API design decisions that I laid out earlier in this thread. Specifically, I am sticking with Jasha's original vision of incorporating this functionality into `just`, `builds`, and `make_config` directly. In the same way that all three of these functions can "auto-config" certain types of data, like complex-numbers, they can now auto-config non-targeted dataclasses. I had been worrying about some potentials issues with integrating dataclass auto-config into `just` et. al, but these ended up proving to be non-issues when I dug further into this. Kudos to @Jasha10 for hitting a bullseye on the design from the get-go! 🥳 


As you can see from the linked WIP PR, I still have some work to do with docs and adding more tests, but please play around with these new capabilities on that branch and see if you see any problems. I wouldn't be surprised if this could get merged this week!

**Comment by rsokl** [comment]
@Jasha10 @addisonklinke @jgbos I have just released `v0.8.0rc3`, which features recursive auto-config of dataclasses via `just`, `builds`, and `make_config` 🥳 !  You can give it a whirl by installing this pre-release: `pip install --pre hydra-zen`

Read more about it here: https://github.com/mit-ll-responsible-ai/hydra-zen/pull/301  

I'm excited about this new feature. I think it is quite powerful!

**Reply by Jasha10** [reply]
Very nice!

**Reply by addisonklinke** [reply]
If I read correctly, the main benefit is being able to maintain more-specific type annotations rather than having them generalized to a Hydra-compatible subset. I could see this being handy down the line as my team's internal code base matures, but to be honest it's not super high on the priority list at the moment. We're mainly focused on auto-generating CLI interfaces for new model/dataset implementations that allow for quick iteration on experimental ideas

**Reply by addisonklinke** [reply]
Actually, would the merge of #301 resolve my FR in #294?

**Reply by rsokl** [reply]
>  the main benefit is being able to maintain more-specific type annotations rather than having them generalized to a Hydra-compatible subset

It also enables default values that hydra-zen provides auto-config support for, like `complex` valued or partial'd functions. As hydra-zen's auto-config support improves, this feature will become more powerful.  Also, the recursive behavior of this is quite powerful -- one design a config with nested dataclasses and call `just` a single time to obtain a Hydra-compatible version of the config. 

I'll also add that the fact that this enables compatibility with pydantic dataclasses (and thus much more general type-checking) is a pretty big deal that I probably need to advertise in our docs more.

**Reply by rsokl** [reply]
> Actually, would the merge of https://github.com/mit-ll-responsible-ai/hydra-zen/pull/301 resolve my FR in https://github.com/mit-ll-responsible-ai/hydra-zen/issues/294?

I wouldn't say so. Here we aren't inferring anything from annotations, we are simply "sanitizing" type annotations. #294 would add the ability for `builds` et al to inspect a target's annotations and to populate the config's signature with sub-configs based on said annotations. I am still interested in adding this feature, but I haven't had time to think carefully about the implementation details and potential issues.

Before tackling #294, I would like to think about the "merge issue" that Jasha nicely summarizes here: https://github.com/mit-ll-responsible-ai/hydra-zen/discussions/293#discussioncomment-3435036

**Reply by addisonklinke** [reply]
> this enables compatibility with pydantic dataclasses
Yes this definitely comes up frequently for me when discussing Hydra-Zen in various Slack communities
